### DISCUSSION TITLE
Use anyio instead of asyncio

### DISCUSSION THREAD

**Discussion by adsko** [sqlalchemy/sqlalchemy#6318] [Ideas]
Currently sqlalchemy requires asyncio in order to work with async programming. 

It would be better to use https://github.com/agronholm/anyio which gives possibility to use different async library like trio.

**Comment by zzzeek** [comment]
hi -

I'm not sure how that works.  asyncpg, which is the main driver everyone is using, [also uses asyncio directly](https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connection.py#L1978).   SQLAlchemy's own use of asyncio is limited to the asyncio.Lock, asyncio.Queue and asyncio.LifoQueue  (asyncpg uses that queue here: https://github.com/MagicStack/asyncpg/blob/master/asyncpg/pool.py#L414, though SQLAlchemy doesn't use this feature).

So what does a library like "trio" do when the whole world is at least modestly hardcoded to asyncio?    adding dependencies to SQLAlchemy is literally something we've never done at all until this year when we added "greenlet", and people are already complaining about that.   So I really don't want to add a dependency here unless the whole world is already standardized on this approach, which does not seem to be used in any of the three asyncio DBAPIs we support (here's [aiomysql](https://github.com/aio-libs/aiomysql/blob/master/aiomysql/connection.py#L478), here's [aiosqlite](https://github.com/omnilib/aiosqlite/blob/main/aiosqlite/core.py#L340))

**Reply by adsko** [reply]
That make sense you dont support different async libraries, while dbs dont support different async implementations.

Trio is still young, but I think the ideas behind handling tasks, cancelations and exceptions are simpler than asyncio.


Let's wait for the future. If asycpg implements different aproaches, it would be nice to come back to this conversation.

**Reply by CaselIT** [reply]
I'm also a bit out of the asyncio world.
So trio is not a different loop, like [uvloop](https://uvloop.readthedocs.io/) that can be used with asyncio, but a different library that uses another api?

**Reply by adsko** [reply]
Yes, Trio has different API, and that's why it wont work with library which uses asyncio lib.

Of course, there are libraries like: https://github.com/python-trio/trio-asyncio which tries to reimplement asyncio loop for trio app, but I'm not sure if it works and how performance looks like.

https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/ Trio owner essay about trio.

**Reply by zzzeek** [reply]
I'm sure trio has all kinds of great ideas but I would ask why aren't these great ideas part of python asyncio?  were they proposed ?     also I thought you can use any event loop you'd like in asyncio, why can't trio's loop be used?  is the issue that you literally can't use asyncio.Lock or asyncio.Queue with trio ?  it seems like trio should find a way to work with that ? (since that's the only explicit asyncio we are using)

**Reply by auvipy** [reply]
instead of using anyio, i would go for trio for future or stick with asyncio deafult.

**Reply by CaselIT** [reply]
I'm still not sure what the advantage would be to sqlalchemy and its users.

All current supported async drivers use asyncio directly, so sqlalchemy would need to continue to support them, meaning we need to use asyncio. Wouldn't just makes more sense to create an sqlalchemy wrapped that exposes the trio api? That must be a common thing since most libraries I know that are async in python use asyncio, not trio.

I admit that I'm far from an expert in async python, so what am I missing?

**Reply by oremanj** [reply]
> why aren't these great ideas part of python asyncio? were they proposed?

Asyncio has been trying to incorporate some of Trio’s core ideas for years. This is difficult due to backwards compatibility considerations. We might finally get asyncio.TaskGroup in 3.11 now that PEP 654 has been approved.

> also I thought you can use any event loop you'd like in asyncio, why can't trio's loop be used?

Trio’s innovations are not at the event loop layer. Substituting a different event loop underneath asyncio wouldn’t improve usability in the way that Trio is trying to do.

> is the issue that you literally can't use asyncio.Lock or asyncio.Queue with trio ? it seems like trio should find a way to work with that ? (since that's the only explicit asyncio we are using)

Yes, that’s the issue. Every async primitive is ultimately implemented in terms of functionality that is specific to the async framework that it runs on. asyncio can’t support trio.Lock either.

——

I agree it probably doesn’t make sense for sqlalchemy to take additional dependencies or go to great lengths to support Trio when all known async db drivers use asyncio. If there db drivers that work on Trio without trio-asyncio, though, it would be nice for sqlalchemy to support them. At that point you could consider adding a customization point for the synchronization primitives that you use. I don’t expect you would need to go all the way to anyio; that’s a big dependency. You could use sniffio (which is a tiny dependency that just provides info about which async library is running) or just let the user choose between asyncio/trio directly (see httpx for an example of how this can work).

**Reply by zzzeek** [reply]
are there any such DB drivers?  what do projects like asyncpg say when this kind of request comes to them?  (after all its main author literally wrote pep 492)?

**Reply by Fuyukai** [reply]
> are there any such DB drivers?

[I wrote one](https://github.com/Fuyukai/pg-purepy).

**Reply by CaselIT** [reply]
> 
> 
> > are there any such DB drivers?
> 
> [I wrote one](https://github.com/Fuyukai/pg-purepy).

Looking at the examples, like here https://pg.py.veriny.tf/en/latest/api-midlevel.html#pg_purepy.connection.AsyncPostgresConnection.fetch_one, it does not seem to require anything different to be used by sqlalchemy. I mean that example looks fairly similar to any other async driver

**Reply by auvipy** [reply]
i dont see much on user side.

**Reply by davidbrochart** [reply]
There was some discussion about using AnyIO in SQLAlchemy [on gitter](https://matrix.to/#/!bYzEZBQEUydwuCVthA:gitter.im/$s4NMfqZ3DYX2yVSANekeiqZzg0wrpbXDu3PdU_mOAPA?via=gitter.im&via=matrix.org&via=matrix.notice.at) recently. According to @graingert it would not require a lot of efforts. There was also a suggestion to use @oremanj's [greenback](https://greenback.readthedocs.io), which I think does what SQLAlchemy does with greenlet, but with support for asyncio and Trio.

**Reply by graingert** [reply]
I think it would be possible to do without a dependency just by using sniffio to pick the trio primative or the asyncio primative (and picking asyncio if sniffio isn't installed) (and caching the sniffio result on the engine)

**Reply by graingert** [reply]
greenback on the other hand is nice because it's a standard approach that works across frameworks *and* libraries. So not only do you get trio+asyncio support one library that works with this sort of greenlet async portal approach will work with another

**Reply by zzzeek** [reply]
as nice as greenback may be, this was an unmaintained library for quite some time and seems to have picked up in the last few years, but overall we're not introducing obscure new dependencies in SQLAlchemy that are not under our control.  Just greenlet alone, which is a very widely used library that's been around for over a decade, has been a huge headache nonetheless.

We have our own third-party greenback-like thing which is called [awaitlet](https://awaitlet.sqlalchemy.org/en/latest/).     it's basically SQLAlchemy's approach ported to its own library.     SQLAlchemy / awaitlet don't depend on each other at all right now, but if we were to lift out SQLAlchemy's greenlet code it would be using awaitlet.   if you want to talk about tweaks to make trio/anyio work in such a way that they remain *optional* dependencies, I would do it over there.

**Reply by davidbrochart** [reply]
Thanks @zzzeek, I opened https://github.com/sqlalchemy/awaitlet/pull/7.
If AnyIO is a library you would be willing to depend on, this should be enough to have `awaitlet` compatible with asyncio and Trio. From what I can see in SQLAlchemy's code base, it should be easy to do the same.
